File: windiskwriter/WimlibWrapper/WimlibWrapper.m
//
//  WimlibWrapper.m
//  windiskwriter
//
//  Created by Macintosh on 12.02.2023.
//  Copyright © 2023 TechUnRestricted. All rights reserved.
//

#import "WimlibWrapper.h"
#import "CChar2DArray.h"
#import "wimlib.h"
#import "wim.h"
#import "xml.h"

@implementation WimlibWrapper {
    WIMStruct *currentWIM;
}

- (instancetype)initWithWimPath: (NSString *)wimPath {
    
    enum wimlib_error_code wimOpenStatus = wimlib_open_wim([wimPath UTF8String], NULL, &currentWIM);
    _wimPath = wimPath;
    
    return self;
}

- (UInt32)imagesCount {
    if (currentWIM == NULL) {
        return 0;
    }
    
    return currentWIM->hdr.image_count;
}

- (WimLibWrapperCPUArch)CPUArchitectureForImageIndex: (UInt32)imageIndex {
    if (currentWIM == NULL) {
        return NULL;
    }
    
    NSString *stringValue = [self propertyValueForKey: @"WINDOWS/ARCH"
                                           imageIndex: imageIndex];
    
    if (stringValue == NULL) {
        return WimLibWrapperCPUArchUnknown;
    }
    
    NSInteger convertedIntegerValue = [stringValue integerValue];
    
    switch (convertedIntegerValue) {
        case WimLibWrapperCPUArchIntel:
        case WimLibWrapperCPUArchMIPS:
        case WimLibWrapperCPUArchAlpha:
        case WimLibWrapperCPUArchPPC:
        case WimLibWrapperCPUArchSHX:
        case WimLibWrapperCPUArchARM:
        case WimLibWrapperCPUArchIA64:
        case WimLibWrapperCPUArchAlpha64:
        case WimLibWrapperCPUArchMSIL:
        case WimLibWrapperCPUArchAMD64:
        case WimLibWrapperCPUArchIA32OnWin64:
        case WimLibWrapperCPUArchARM64:
            return convertedIntegerValue;
        default:
            return WimLibWrapperCPUArchUnknown;
    }
}

- (NSString *_Nullable)propertyValueForKey: (NSString *)key
                                imageIndex: (NSUInteger)imageIndex {
    if (currentWIM == NULL) {
        return NULL;
    }
    
    return [NSString stringWithCString: wimlib_get_image_property(currentWIM, imageIndex, key.UTF8String)
                              encoding: NSUTF8StringEncoding];
}

- (WimlibWrapperResult)setPropertyValue: (NSString *)value
                                 forKey: (NSString *)key
                             imageIndex: (UInt32)imageIndex {
    if (currentWIM == NULL) {
        return NO;
    }
    
    NSString *currentValue = [self propertyValueForKey: key
                                            imageIndex: imageIndex];
    
    if ([value isEqualToString:currentValue]) {
        return WimlibWrapperResultSkipped;
    }
    
    enum wimlib_error_code result = wimlib_set_image_property(currentWIM,
                                                              imageIndex,
                                                              [key cStringUsingEncoding: NSUTF8StringEncoding],
                                                              [value cStringUsingEncoding: NSUTF8StringEncoding]);
    
    if (result != WIMLIB_ERR_SUCCESS) {
        return WimlibWrapperResultFailure;
    }
    
    return WimlibWrapperResultSuccess;
}

- (WimlibWrapperResult)setPropertyValueForAllImages: (NSString *)value
                                             forKey: (NSString *)key {
    
    UInt32 imagesCount = [self imagesCount];
    
    BOOL requiresOverwriting = NO;
    
    for (UInt32 currentImageIndex = 1; currentImageIndex <= imagesCount; currentImageIndex++) {
        WimlibWrapperResult setPropertyResult = [self setPropertyValue: value
                                                                forKey: key
                                                            imageIndex: currentImageIndex];
        
        switch (setPropertyResult) {
            case WimlibWrapperResultSuccess:
                requiresOverwriting = YES;
                break;
            case WimlibWrapperResultFailure:
                return WimlibWrapperResultFailure;
            case WimlibWrapperResultSkipped:
                break;
        }
    }
    
    return requiresOverwriting ? WimlibWrapperResultSuccess : WimlibWrapperResultSkipped;
}

- (BOOL)applyChanges {
    if (currentWIM == NULL) {
        return NO;
    }
    
    enum wimlib_error_code overwriteReturnCode = wimlib_overwrite(currentWIM, 0, 1);
    
    return overwriteReturnCode == WIMLIB_ERR_SUCCESS;
}

- (enum wimlib_error_code)splitWithDestinationDirectoryPath: (NSString *)destinationDirectoryPath
                                        maxSliceSizeInBytes: (UInt64 *)maxSliceSizeInBytes
                                            progressHandler: (wimlib_progress_func_t _Nullable)progressHandler
                                                    context: (void *_Nullable)context {
    if (currentWIM == NULL) {
        return WIMLIB_ERR_ABORTED_BY_PROGRESS;
    }
    
    if (progressHandler != NULL) {
        wimlib_register_progress_function(currentWIM, progressHandler, context);
    }
    
    NSString *destinationFileName = [[[_wimPath lastPathComponent] stringByDeletingPathExtension] stringByAppendingPathExtension:@"swm"];
    
    return wimlib_split(currentWIM,
                        [[destinationDirectoryPath stringByAppendingPathComponent:destinationFileName] UTF8String],
                        maxSliceSizeInBytes,
                        NULL
                        );
}

- (BOOL)extractFiles: (NSArray *)files
destinationDirectory: (NSString *)destinationDirectory
      fromImageIndex: (UInt32)imageIndex {
    
    if (currentWIM == NULL) {
        return NULL;
    }
    
    CChar2DArray *filesArrayCCharEncoded = [[CChar2DArray alloc] initWithNSArray:files];
    
    enum wimlib_error_code extractionResult = wimlib_extract_paths(currentWIM,
                                                                   imageIndex,
                                                                   [destinationDirectory UTF8String],
                                                                   [filesArrayCCharEncoded getArray],
                                                                   [files count],
                                                                   WIMLIB_EXTRACT_FLAG_NO_PRESERVE_DIR_STRUCTURE
                                                                   );
    
    return extractionResult == WIMLIB_ERR_SUCCESS;
}

- (WimlibWrapperResult)extractWindowsEFIBootloaderForDestinationDirectory: (NSString *)destinationDirectory {
    UInt32 imagesCount = [self imagesCount];
    
    for (UInt32 currentImageIndex = 1; currentImageIndex <= imagesCount; currentImageIndex++) {
        if ([self CPUArchitectureForImageIndex:currentImageIndex] != WimLibWrapperCPUArchAMD64) {
            continue;
        }
        
        BOOL bootloaderExtractionResult = [self extractFiles: @[@"/Windows/Boot/EFI/bootmgfw.efi"]
                                        destinationDirectory: destinationDirectory
                                              fromImageIndex: currentImageIndex];
        
        if (!bootloaderExtractionResult) {
            return WimlibWrapperResultFailure;
        }
        
        BOOL bootloaderRanamingSuccess = [NSFileManager.defaultManager moveItemAtPath: [destinationDirectory stringByAppendingPathComponent: @"bootmgfw.efi"]
                                                                               toPath: [destinationDirectory stringByAppendingPathComponent: @"bootx64.efi"]
                                                                                error: NULL];
        if (!bootloaderRanamingSuccess) {
            return NO;
        }
        
        return WimlibWrapperResultSuccess;
    }
    
    return WimlibWrapperResultSkipped;
}

- (WimlibWrapperResult)patchWindowsRequirementsChecks {
    if (currentWIM == NULL) {
        return WimlibWrapperResultFailure;
    }
    
    WimlibWrapperResult setPropertyResult = [self setPropertyValueForAllImages: @"Server"
                                                                        forKey: @"WINDOWS/INSTALLATIONTYPE"];
    
    switch (setPropertyResult) {
        case WimlibWrapperResultSkipped:
        case WimlibWrapperResultFailure:
            return setPropertyResult;
        default:
            break;
    }
    
    WimlibWrapperResult applyChangesResult = [self applyChanges];
    
    return applyChangesResult;
}

- (void)dealloc {
    if (currentWIM != NULL) {
        wimlib_free(currentWIM);
    }
}

@end

File: WinDiskWriter/Services/WimlibWrapper/WimlibWrapper.swift
//
//  WimlibWrapper.swift
//  WinDiskWriter
//
//  Created by Macintosh on 07.07.2024.
//

import Foundation

fileprivate extension Int32 {
    var asWimlibReturnCode: wimlib_error_code {
        return wimlib_error_code(UInt32(self))
    }
}

// MARK: - Wrapper Main Components
final class WimlibWrapper {
    private var currentWIM: UnsafeMutablePointer<WIMStruct>?
    private let wimURL: URL

    var imagesCount: Int32 {
        guard let currentWIM = currentWIM else {
            return 0
        }

        return Int32(currentWIM.pointee.hdr.image_count)
    }

    init?(with wimURL: URL) {
        self.wimURL = wimURL

        let wimPath: String = wimURL.path
        let openFlags: Int32 = 0

        let wimOpenStatus = wimlib_open_wim(wimPath, openFlags, &currentWIM).asWimlibReturnCode

        if wimOpenStatus != WIMLIB_ERR_SUCCESS {
            return nil
        }
    }

    deinit {
        if let currentWIM = currentWIM {
            wimlib_free(currentWIM)
        }
    }

    func applyChanges() -> Bool {
        guard let currentWIM = currentWIM else {
            return false
        }

        let writeFlags: Int32 = 0
        let numberOfThreads: UInt32 = 1

        let result = wimlib_overwrite(currentWIM, writeFlags, numberOfThreads).asWimlibReturnCode

        return result == WIMLIB_ERR_SUCCESS
    }
}

// MARK: - Files Manipulation
extension WimlibWrapper {
    func extractFiles(_ files: [String], to destination: String, fromImageIndex index: Int32) -> Bool {
        guard let currentWIM = currentWIM else {
            return false
        }

        var wimlibResult: wimlib_error_code = WIMLIB_ERR_INVALID_PARAM

        files.charUnsafePointerArray { charPointerArray in
            let result = wimlib_extract_paths(
                currentWIM,
                index,
                destination,
                charPointerArray,
                files.count,
                WIMLIB_EXTRACT_FLAG_NO_PRESERVE_DIR_STRUCTURE
            )

            wimlibResult = result.asWimlibReturnCode
        }

        return wimlibResult == WIMLIB_ERR_SUCCESS
    }
}

// MARK: - Image Property Processor
extension WimlibWrapper {
    func propertyValue(forKey key: String, imageIndex: Int32) -> String? {
        guard let currentWIM = currentWIM else {
            return nil
        }

        guard let value = wimlib_get_image_property(currentWIM, imageIndex, key) else {
            return nil
        }

        return String(cString: value)
    }

    func setPropertyValue(_ value: String, forKey key: String, imageIndex: Int32) -> WimlibWrapperResult {
        guard let currentWIM = currentWIM else {
            return .failure
        }

        guard let currentValue = propertyValue(forKey: key, imageIndex: imageIndex), currentValue != value else {
            return .skipped
        }

        let result = wimlib_set_image_property(currentWIM, imageIndex, key, value).asWimlibReturnCode

        return result == WIMLIB_ERR_SUCCESS ? .success : .failure
    }

    func setPropertyValueForAllImages(_ value: String, forKey key: String) -> WimlibWrapperResult {
        var requiresOverwriting: Bool = false

        for index in 1...imagesCount {
            let result = setPropertyValue(value, forKey: key, imageIndex: index)
            switch result {
            case .success:
                requiresOverwriting = true
            case .failure:
                return .failure
            case .skipped:
                continue
            }
        }

        return requiresOverwriting ? .success : .skipped
    }
}
